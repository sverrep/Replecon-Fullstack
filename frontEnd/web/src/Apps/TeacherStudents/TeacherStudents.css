.teacher-navbar{
    background-color: #546F99;
}

.wrapper{
    font-family: 'Montserrat', sans-serif;
}

.student-container{
    align-items: center;
    justify-content: center;
    height: 100%;
    padding-top: 3%;
    font-family: 'Montserrat', sans-serif;
    overflow: visible;
}

.student-header{
    text-align: center;
}

.student-list-col{
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    align-items: center;
}

.list-group{
    max-height: 80%;
    width: 70%;
    overflow-y:scroll;
    -webkit-overflow-scrolling: touch;
}

.list-group-item.active{
    background-color: #546F99 !important;
}

.pay-btns {
    margin-top: 2%;
    width: 60%;
    margin-right: 5px;
    margin-left: 5px;
    background-color: #546F99 !important;
    border-color: #546F99 !important;
    border-radius: 12px !important;
}

.pay-btns:hover {
    background-color: #3a4d6b !important;
}

.amount-input{
    margin-top: 2%;
    width: 80%;
}

.list-balance-col{
    display: flex;
    justify-content: flex-end;
}