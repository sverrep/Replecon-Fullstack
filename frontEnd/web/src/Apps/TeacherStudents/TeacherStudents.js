import React from 'react';
import { withRouter } from "react-router-dom";
import axios from 'axios';
import getIP from '../../settings.js';
import './TeacherStudents.css';
import navbar from '../../Components/navbar/Teacher NavBar/TeacherNavbar';
import Button from 'react-bootstrap/Button'
import ListGroup from 'react-bootstrap/ListGroup'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import InputGroup from 'react-bootstrap/InputGroup'
import FormControl from 'react-bootstrap/FormControl'
import Cookies from 'universal-cookie';



class TeacherStudents extends React.Component {
    constructor(props) {
        super(props);
        this.state = { 
            class_code: this.props.location.state.class.class_code, 
            class_name: this.props.location.state.class.class_name,
            teacher_id:  this.props.location.state.teacher_id,
            students: [],
            selected: [],
            selected_name: "",
            selected_balance: "",
            selected_id: "",
            amount: "",
            display_items: [],
            all_items: [],
        };
        this.studentClicked = this.studentClicked.bind(this)
        this.renderStudents = this.renderStudents.bind(this)
        this.handleChange = this.handleChange.bind(this)
        this.amountIsValid = this.amountIsValid.bind(this)
    }

    async componentDidMount(){
        const cookies = new Cookies()
        axios.defaults.headers.common.Authorization = cookies.get("Authorization");
        await this.getClassStudents()
        await this.getAllItems()
        
    }

    async getClassStudents() {
        await axios.get(getIP()+'/students/')
        .then(response => {
            this.getCurrentClassStudents(response.data)
        })
        .catch(error => this.setState({error: error}))
    }

    async getAllItems(){
        await axios.get(getIP()+'/items/')
        .then(async (response) => {
            this.setState({all_items: response.data})
        })
        .catch(error => this.setState({error: error}))
    }

    handleChange(e) {
        const field = e.target.id
        this.setState({ [field] : e.target.value });
    }

    getCurrentClassStudents(students){
        var newAr = []
        for (let i = 0; i<=Object.keys(students).length -1;i++)
        {
            if(students[i].class_code===this.state.class_code)
            {
              newAr.push(students[i])
            }
        }
        this.setState({students:newAr})
    }

    studentClicked(student){
        this.setState({selected_name: student.name})
        this.setState({selected_balance: student.balance})
        this.setState({selected_id: student.id})
        var select = false;
        var index = 0;
        for(let i = 0; i <= Object.keys(this.state.selected).length-1; i++)
        {
          if(student.id === this.state.selected[i])
          {
              select = true;
              index = i;
              break;
          }
        }
        if(select === true)
        {
          var temparray = this.state.selected
          temparray.splice(index, 1)
          this.setState({selected: temparray})
        }
        else
        {
          this.setState(prevState => ({selected: [...prevState.selected, student.id]}))
          this.getBoughtItems()
        }
    }

    renderStudents(student, x){
        var select = false;
        for(let i = 0; i <= Object.keys(this.state.selected).length-1; i++)
        {
            if(student.id === this.state.selected[i])
            {
                select = true;
                break;
            }
        }
        if (select === true)
        {
            return(
                <ListGroup.Item key={x} active action onClick={() => this.studentClicked(student)}>
                    <Row>
                        <Col>
                            {student.name}
                        </Col>
                        <Col  className="list-balance-col">
                            {"Balance: " + student.balance}
                        </Col>
                    </Row>
                </ListGroup.Item>
            )
        }
        else if (select === false)
        {
            return(
                <ListGroup.Item key={x} action onClick={() => this.studentClicked(student)}>
                    <Row>
                        <Col>
                            {student.name}
                        </Col>
                        <Col  className="list-balance-col">
                            {"Balance: " + student.balance}
                        </Col>
                    </Row>
                </ListGroup.Item>
            )
        }
    }

    async changeSelected(pay){
        var selected_amount = this.state.amount 
        if(this.amountIsValid(selected_amount)) {
          if(!pay) {
            selected_amount= "-" + this.state.amount
          }
          var selected = this.state.selected
          for(let i = 0; i <= Object.keys(selected).length-1; i++)
          {
            var payload = { user_id: selected[i], amount: selected_amount }; 
            await axios.put(getIP()+'/students/balance/', payload)
            .then(async response => {
                await axios.post(getIP()+'/transactions/teacherPayStudents/', {"user_id": response.data.user, "amount": selected_amount})
                .then(async response => {
                  await this.getClassStudents()
                  this.setState({selected: []})
                })
                .catch(error => console.log(error))
            })
            .catch(error => this.setState({error: error}))
          }
          this.setState({show:false})
        }
        else
        {
            this.setState({error: "Did not pass validation"})
        }
    }

    async getBoughtItems(){
        await axios.get(getIP()+'/items/allboughtitems/')
        .then(response => {
            var all_bought_items = response.data.reverse()
            var temparray = []
            for(var i = 0; i < Object.keys(all_bought_items).length; i++)
            {
                for(var j = 0; j < Object.keys(this.state.selected).length; j++)
                {
                    
                    if (this.state.selected[j] === all_bought_items[i].user_id)
                    {
                        temparray.push(all_bought_items[i])
                    }
                }
            }
            this.setState({display_items: temparray})
        })
        .catch(error => this.setState({error: error}))
    }

    renderItems(item, x){
        var name = ""
        var item_name = ""
        var selected = false
        for(var i = 0; i < Object.keys(this.state.students).length; i++)
        {
            if(item.user_id === this.state.students[i].id)
            {
                name = this.state.students[i].name
            }
        }
        for (var j = 0; j < Object.keys(this.state.all_items).length; j++)
        {
            if(item.item_id === this.state.all_items[j].id)
            {
                item_name = this.state.all_items[j].item_name
            }
        }
        for (var k = 0; k < Object.keys(this.state.selected).length; k++)
        {
            if(item.user_id === this.state.selected[k])
            {
                selected = true
            }
        }
        if(selected === true)
        {
            return(
                <ListGroup.Item key={x}>
                    <Row>
                        <Col>
                            {item_name}
                        </Col>
                        <Col  className="list-balance-col">
                            {name}
                        </Col>
                    </Row>
                </ListGroup.Item>
            );
        }
    }

    async deleteSelected(){
        for (let i = 0; i < Object.keys(this.state.selected).length; i++)
        {
            await axios.delete(getIP()+`/users/${this.state.selected[i]}/`)
            .then(async response => {
                await axios.get(getIP()+`/transactions/getTransactionsByID/${this.state.selected[i]}`)
                .then(async response1 => {
                    for (var j = 0; j < Object.keys(response1.data).length; j++)
                    {
                        if (response1.data[j].name === "BANK")
                        {
                            await axios.delete(getIP()+`/transactioninterestrates/${response1.data[j].id}`)
                            .then(async response2 => {
                            })
                            .catch(error => console.log(error))
                        }
                    }
                })
                this.getClassStudents()
            })
            .catch(error => console.log(error))
        }
    }

    amountIsValid(amount){
        if(isNaN(amount)){
          this.setState({error: 'Make sure that amount is a number'})
          return false
        }
        else{
          if(Math.sign(amount) === 1){
              return true
          }
          else{
              this.setState({error: 'Make sure that amount is a positive number'})
              return false
          }
        }
    } 

    render(){
        return (
            <div className="wrapper">
                {navbar(this.state.class_name, {class_code: this.state.class_code, class_name: this.state.class_name}, this.state.teacher_id)}
                <Container className="student-container">
                    <Row className="content-row">
                        <Col className="student-list-col">
                            <h4 className="student-header">Students</h4>
                            <ListGroup>
                                {this.state.students.map((student,i) => this.renderStudents(student, i))}
                            </ListGroup>
                            <InputGroup className="amount-input">
                                <InputGroup.Text>$</InputGroup.Text>
                                <FormControl defaultValue="0" id='amount' onChange={this.handleChange} />
                            </InputGroup>
                            <p>{this.state.error}</p>
                            <Button className="pay-btns" onClick={() => this.changeSelected(true)}>Pay Selected Students</Button>
                            <Button className="pay-btns" onClick={() => this.changeSelected(false)}>Charge Selected Students</Button>
                            <Button className="delete-btn" onClick={() => this.deleteSelected()}>Delete Selected Students</Button>
                        </Col>
                        <Col>
                            <h4 className="student-header">Student Items</h4>
                            {this.state.display_items.map((item,i) => this.renderItems(item, i))}
                        </Col>
                    </Row>
                </Container>
            </div>
        );
    }
}

export default withRouter(TeacherStudents)