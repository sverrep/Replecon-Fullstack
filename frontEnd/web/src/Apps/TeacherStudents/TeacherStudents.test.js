import React from 'react';
import TeacherStudents from './TeacherStudents.js'
import axios from 'axios'
jest.mock('axios');
let wrapper

describe('TeacherStudents Testing', () => {
    beforeEach(() => {
        const props = {
            location: {
                state: {
                    class: {
                        class_code: "classc",
                        class_name: "classname"
                    },
                    teacher_id: "teacherid",
                    token: "token",
                }
            }
        }
        wrapper = shallow(<TeacherStudents.WrappedComponent {...props} />, { disableLifecycleMethods: true })
    })
    describe('componentMount testing', () => {
        it('should call the right functions', async () => {
            const instance = wrapper.instance()

            instance.getClassStudents = jest.fn()
            instance.getAllItems = jest.fn()

            await instance.componentDidMount()

            expect(instance.getClassStudents).toHaveBeenCalledTimes(1)
            expect(instance.getAllItems).toHaveBeenCalledTimes(1)
        });

        it('should pass', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const items = mockItemData()
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))
            axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.componentDidMount()
            expect(instance.state.students).toEqual(students.data)
            expect(instance.state.all_items).toEqual(items.data)
        });   
    })
    describe('getClassStudents testing', () => {
        it('should reject axios call', async () => {
            const instance = wrapper.instance()

            axios.get.mockImplementationOnce(() => Promise.reject("Token not authenticated"))
            await instance.getClassStudents()
            expect(instance.state.error).toBe("Token not authenticated")
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/')

        });

        it('should pass axios call and call getCurrentClassStudents', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))
            instance.getCurrentClassStudents = jest.fn()

            await instance.getClassStudents()
            expect(instance.getCurrentClassStudents).toHaveBeenCalledTimes(1)
            expect(instance.getCurrentClassStudents).toHaveBeenCalledWith(students.data)
        }); 

        it('should pass axios call and set students state', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))

            await instance.getClassStudents()
            expect(instance.state.students).toEqual(students.data)
        }); 
    })
    describe('getAllItems testing', () => {
        it('should reject axios call', async () => {
            const instance = wrapper.instance()

            axios.get.mockImplementationOnce(() => Promise.reject("Token not authenticated"))
            await instance.getAllItems()
            expect(instance.state.error).toBe("Token not authenticated")
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//items/')

        });

        it('should pass axios call and call set correct state', async () => {
            const instance = wrapper.instance()
            const items = mockItemData()
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.getAllItems()
            expect(instance.state.all_items).toEqual(items.data)
        }); 
    })
    describe('getCurrentClassStudents testing', () => {
        it('should find no students', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            students.data[0].class_code = "classa"
            students.data[1].class_code = "classb"
            instance.getCurrentClassStudents(students.data)
            expect(instance.state.students).toEqual([])

        });
        it('should find 1 student and write it to state', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            students.data[1].class_code = "classb"
            instance.getCurrentClassStudents(students.data)
            expect(instance.state.students).toHaveLength(1)
            expect(instance.state.students[0]).toEqual(students.data[0])
        });
        it('should find both student and write it to state', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            instance.getCurrentClassStudents(students.data)
            expect(instance.state.students).toHaveLength(2)
            expect(instance.state.students).toEqual(students.data)
        });
    })
    describe('studentClicked testing', () => {
        it('should add student to selected and call getBoughtItems', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const student = students.data[0]

            instance.getBoughtItems = jest.fn()

            instance.studentClicked(student)
            expect(instance.state.selected_name).toEqual(student.name)
            expect(instance.state.selected_balance).toEqual(student.balance)
            expect(instance.state.selected_id).toEqual(student.id)
            expect(instance.state.selected).toHaveLength(1)
            expect(instance.state.selected[0]).toEqual(student.id)
            expect(instance.getBoughtItems).toHaveBeenCalledTimes(1)

        });

        it('should find student in selected and remove them from it', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const student = students.data[0]
            instance.state.selected[0] = student.id

            instance.getBoughtItems = jest.fn()

            instance.studentClicked(student)
            expect(instance.state.selected_name).toEqual(student.name)
            expect(instance.state.selected_balance).toEqual(student.balance)
            expect(instance.state.selected_id).toEqual(student.id)
            expect(instance.state.selected).toHaveLength(0)
            expect(instance.getBoughtItems).toHaveBeenCalledTimes(0)

        });

        it('should find student in selected and remove them from it', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const student = students.data[0]
            const all_bought_items = mockBoughtItems()

            axios.get.mockImplementationOnce(() =>  Promise.resolve(all_bought_items))

            instance.studentClicked(student)

            expect(instance.state.selected_name).toEqual(student.name)
            expect(instance.state.selected_balance).toEqual(student.balance)
            expect(instance.state.selected_id).toEqual(student.id)
            expect(instance.state.selected).toHaveLength(1)
            expect(instance.state.selected[0]).toEqual(student.id)
            expect(instance.state.display_items)

        });

    })
    describe('renderStudents testing', () => {
        it('should display not selected student', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const student = students.data[0]

            expect(instance.renderStudents(student)).toMatchSnapshot();

        });

        it('should display a selected student', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const student = students.data[0]
            instance.state.selected[0] = student.id
            expect(instance.renderStudents(student)).toMatchSnapshot();
        });


    })
    describe('changeSelected testing', () => {
        it('should not pass amountIsValid', async () => {
            const instance = wrapper.instance()
            instance.amountIsValid = jest.fn()
            instance.amountIsValid.mockReturnValueOnce(false)
            await instance.changeSelected()
            expect(instance.state.error).toEqual("Did not pass validation")
        });

        it('should not pass through loop', async () => {
            const instance = wrapper.instance()
            instance.amountIsValid = jest.fn()
            instance.amountIsValid.mockReturnValueOnce(true)
            instance.changeSelected()
            expect(axios.put).toHaveBeenCalledTimes(0)
            expect(instance.state.show).toBe(false)
        });

        it('should not pass through axios', async () => {
            const instance = wrapper.instance()
            instance.amountIsValid = jest.fn()
            instance.amountIsValid.mockReturnValueOnce(true)
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id 

            axios.put.mockImplementationOnce(() => Promise.reject("Token not authenticated"))

            await instance.changeSelected()
            expect(axios.put).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledTimes(0)
            expect(instance.state.error).toEqual("Token not authenticated")
        });

        it('should pass axios and call getClassStudents with negative amount', async () => {
            const instance = wrapper.instance()
            instance.amountIsValid = jest.fn()
            instance.amountIsValid.mockReturnValueOnce(true)
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id 
            instance.state.amount = 10
            instance.getClassStudents = jest.fn()
            const user_id = {data: {user: "studentuserid"}}

            axios.put.mockImplementationOnce(() => Promise.resolve(user_id))
            axios.post.mockImplementationOnce(() => Promise.resolve())

            await instance.changeSelected()

            expect(axios.put).toHaveBeenCalledTimes(1)
            expect(axios.put).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//students/balance/", {"amount": "-10", "user_id": students.data[0].id})
            expect(axios.post).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//transactions/teacherPayStudents/",{"amount": "-10", "user_id": user_id.data.user})
            expect(instance.getClassStudents).toHaveBeenCalledTimes(1)
            expect(instance.state.selected).toHaveLength(0)
        });

        it('should pass axios and call getClassStudents with positive amount', async () => {
            const instance = wrapper.instance()
            instance.amountIsValid = jest.fn()
            instance.amountIsValid.mockReturnValueOnce(true)
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id 
            instance.state.amount = 10
            instance.getClassStudents = jest.fn()
            const user_id = {data: {user: "studentuserid"}}

            axios.put.mockImplementationOnce(() => Promise.resolve(user_id))
            axios.post.mockImplementationOnce(() => Promise.resolve())

            await instance.changeSelected(true)

            expect(axios.put).toHaveBeenCalledTimes(1)
            expect(axios.put).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//students/balance/", {"amount": 10, "user_id": students.data[0].id})
            expect(axios.post).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//transactions/teacherPayStudents/",{"amount": 10, "user_id": user_id.data.user})
            expect(instance.getClassStudents).toHaveBeenCalledTimes(1)
            expect(instance.state.selected).toHaveLength(0)
        });

        it('should pass', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id 
            instance.state.amount = 10
            const user_id = {data: {user: "studentuserid"}}

            axios.put.mockImplementationOnce(() => Promise.resolve(user_id))
            axios.post.mockImplementationOnce(() => Promise.resolve())
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))
            await instance.changeSelected(true)

            expect(axios.put).toHaveBeenCalledTimes(1)
            expect(axios.put).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//students/balance/", {"amount": 10, "user_id": students.data[0].id})
            expect(axios.post).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//transactions/teacherPayStudents/",{"amount": 10, "user_id": user_id.data.user})
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/')
            expect(instance.state.selected).toHaveLength(0)
            expect(instance.state.students).toEqual(students.data)
        });
    })
    describe('getBoughtItems testing', () => {
        it('should reject axios call', async () => {
            const instance = wrapper.instance()

            axios.get.mockImplementationOnce(() => Promise.reject("Token not authenticated"))
            await instance.getBoughtItems()
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.get).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//items/allboughtitems/")
            expect(instance.state.error).toEqual("Token not authenticated")

        });

        it('should pass axios call and find no bought items', async () => {
            const instance = wrapper.instance()
            const items = mockBoughtItems()
            const students = mockStudentData()
            instance.state.selected[0] = students.data[1].id
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.getBoughtItems()
            expect(instance.state.display_items).toHaveLength(0)
        }); 

        it('should pass axios call and find set bought items to display state', async () => {
            const instance = wrapper.instance()
            const items = mockBoughtItems()
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id
            
            axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.getBoughtItems()
            expect(instance.state.display_items).toHaveLength(2)
            expect(instance.state.display_items).toEqual(items.data)
        });
    })
    describe('renderItems testing', () => {
        it('should not display not selected students', () => {
            const instance = wrapper.instance()
            const items = mockItemData()
            const item = items.data[0]

            expect(instance.renderItems(item)).toMatchSnapshot();

        });

        it('should display selected students items', () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            const boughtitems = mockBoughtItems()
            const items = mockItemData()
            const student = students.data[0]
            const boughtitem = boughtitems.data[0]
            instance.state.selected[0] = student.id
            instance.state.students[0] = student
            instance.state.all_items[0] = items.data[0]
            expect(instance.renderItems(boughtitem)).toMatchSnapshot();
        });


    })
    describe('deleteSelected testing', () => {
        it('should not find any selected', async () => {
            const instance = wrapper.instance()
            await instance.deleteSelected()
            expect(axios.delete).toHaveBeenCalledTimes(0)
        });

        it('should pass', async () => {
            const instance = wrapper.instance()
            const students = mockStudentData()
            instance.state.selected[0] = students.data[0].id 
            const user_id = {data: [{name: "BANK", id: 'bankid'}]}

            axios.delete.mockImplementationOnce(() => Promise.resolve())
            axios.get.mockImplementationOnce(() => Promise.resolve(user_id))
            axios.delete.mockImplementationOnce(() =>  Promise.resolve())
            axios.get.mockImplementationOnce(() =>  Promise.resolve(students))
            await instance.deleteSelected(true)
            expect(axios.delete).toHaveBeenCalledTimes(2)
            expect(axios.delete).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//users/studentid/")
            expect(axios.delete).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//transactioninterestrates/bankid")
            expect(axios.get).toHaveBeenCalledTimes(2)
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//transactions/getTransactionsByID/studentid')
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/')
            expect(instance.state.students).toEqual(students.data)
        });
    })
    describe('amountIsValid testing', () => {
        it('should fail amountIsValid due to amount not being number', () => {
            const instance = wrapper.instance()
            const condition = instance.amountIsValid("yes")
            expect(instance.state.error).toBe("Make sure that amount is a number")
            expect(condition).toBe(false)
        });
        
        it('should fail amountIsValid due to negative number', () => {
            const instance = wrapper.instance()
            const condition = instance.amountIsValid(-10)
            expect(instance.state.error).toBe("Make sure that amount is a positive number")
            expect(condition).toBe(false)
        });

        it('should pass', () => {
            const instance = wrapper.instance()
            const condition = instance.amountIsValid(10)
            expect(instance.state.error).toBe(undefined)
            expect(condition).toBe(true)
        });
    })
    describe('component render testing', () => {
        it('TeacherStudents page renders default correctly', async () => {
            expect(wrapper.getElements()).toMatchSnapshot();
        })
        
    })

})

const mockStudentData = jest.fn(() => {
    return {
        data: [{id: "studentid", name: "studentname", class_code: "classc", balance: "balance"}, {id: "2studentid", name: "2studentname", class_code: "classc", balance: "2balance"}],
    };
});

const mockItemData = jest.fn(() => {
    return {
        data: [{id: "itemid", item_name: "itemname", description: "itemdesc", price: "itemprice", shop_id: "itemshopid"}, {id: "2itemid", item_name: "2itemname", description: "2itemdesc", price: "2itemprice", shop_id: "2itemshopid"}]
    };
});

const mockBoughtItems = jest.fn(() => {
    return {
        data: [{item_id: "itemid", user_id: "studentid"}, {item_id: "2itemid", user_id: "studentid"}]
    };
});


//{"id": student.user.id, "name": student.user.first_name, "class_code": student.class_code, "balance": student.balance}
//fields = ['id', 'item_name', 'description', 'price', 'shop_id']
//fields = ['item_id', 'user_id']