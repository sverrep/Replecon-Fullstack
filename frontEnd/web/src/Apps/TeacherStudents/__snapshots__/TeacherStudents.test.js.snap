// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeacherStudents Testing component render testing TeacherStudents page renders default correctly 1`] = `
Array [
  <div
    className="wrapper"
  >
    <Navbar
      className="teacher-navbar"
      collapseOnSelect={true}
      expand="lg"
      variant="dark"
    >
      <Container
        fluid={false}
      >
        <NavbarBrand
          className="navbar-header"
        >
          classname
        </NavbarBrand>
        <Nav
          className="navbar-btns"
          fill={false}
          justify={false}
        >
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Class/classc/students",
                  "state": Object {
                    "class": Object {
                      "class_code": "classc",
                      "class_name": "classname",
                    },
                    "teacher_id": "teacherid",
                    "token": "token",
                  },
                }
              }
            >
              Students
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Class/classc/taxes",
                  "state": Object {
                    "class": Object {
                      "class_code": "classc",
                      "class_name": "classname",
                    },
                    "teacher_id": "teacherid",
                    "token": "token",
                  },
                }
              }
            >
              Taxes
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Class/classc/store",
                  "state": Object {
                    "class": Object {
                      "class_code": "classc",
                      "class_name": "classname",
                    },
                    "teacher_id": "teacherid",
                    "token": "token",
                  },
                }
              }
            >
              Store
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Class/classc/bank",
                  "state": Object {
                    "class": Object {
                      "class_code": "classc",
                      "class_name": "classname",
                    },
                    "teacher_id": "teacherid",
                    "token": "token",
                  },
                }
              }
            >
              Bank
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn-class-list"
            disabled={false}
            variant="secondary-outline"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Profile/",
                  "state": Object {
                    "role": "Teacher",
                    "token": "token",
                  },
                }
              }
            >
              Class List
            </Link>
          </Button>
        </Nav>
      </Container>
    </Navbar>
    <Container
      className="student-container"
      fluid={false}
    >
      <Row
        className="content-row"
      >
        <Col
          className="student-list-col"
        >
          <h4
            className="student-header"
          >
            Students
          </h4>
          <ListGroup />
          <InputGroup
            className="amount-input"
          >
            <InputGroupText>
              $
            </InputGroupText>
            <FormControl
              defaultValue="0"
              id="amount"
              onChange={[Function]}
            />
          </InputGroup>
          <p />
          <Button
            active={false}
            className="pay-btns"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
            Pay Selected Students
          </Button>
          <Button
            active={false}
            className="pay-btns"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
            Charge Selected Students
          </Button>
          <Button
            active={false}
            className="delete-btn"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
            Delete Selected Students
          </Button>
        </Col>
        <Col>
          <h4
            className="student-header"
          >
            Student Items
          </h4>
        </Col>
      </Row>
    </Container>
  </div>,
]
`;

exports[`TeacherStudents Testing renderItems testing should display selected students items 1`] = `
<ListGroupItem>
  <Row>
    <Col>
      itemname
    </Col>
    <Col
      className="list-balance-col"
    >
      studentname
    </Col>
  </Row>
</ListGroupItem>
`;

exports[`TeacherStudents Testing renderItems testing should not display not selected students 1`] = `undefined`;

exports[`TeacherStudents Testing renderStudents testing should display a selected student 1`] = `
<ListGroupItem
  action={true}
  active={true}
  onClick={[Function]}
>
  <Row>
    <Col>
      studentname
    </Col>
    <Col
      className="list-balance-col"
    >
      Balance: balance
    </Col>
  </Row>
</ListGroupItem>
`;

exports[`TeacherStudents Testing renderStudents testing should display not selected student 1`] = `
<ListGroupItem
  action={true}
  onClick={[Function]}
>
  <Row>
    <Col>
      studentname
    </Col>
    <Col
      className="list-balance-col"
    >
      Balance: balance
    </Col>
  </Row>
</ListGroupItem>
`;
